ddht/v5_1/alexandria/partials/proof.py

```diff
@@ -80,6 +164,195 @@ def get_tree(self) -> ProofTree:
+        )
+        expected_partial_chunk_count = (
+            last_partial_chunk_index - first_partial_chunk_index
+        ) + 1
+        if len(leaf_nodes_for_partial) != expected_partial_chunk_count:
+            raise Exception(
+                "Proof is missing leaf nodes required for partial construction."
+            )
+
+        # Compute the total number of non-padding chunks in the tree.
+        num_data_chunks = get_chunk_count_for_data_length(length)
+        last_data_chunk_index = max(0, num_data_chunks - 1)
+        last_data_chunk_path = chunk_index_to_path(
+            last_data_chunk_index, self.path_bit_length
+        )
+
+        #
+        # Data chunks left of the partial
+        #
+        # start-at : 0
+        # end-at   : first_partial_chunk_index - 1
+        if first_partial_chunk_index > 0:
+            nodes_left_of_partial = tree.get_subtree_proof_nodes(
+                chunk_index_to_path(0, self.path_bit_length),
+                chunk_index_to_path(
+                    first_partial_chunk_index - 1, self.path_bit_length
+                ),
+            )
+        else:
+            nodes_left_of_partial = ()
+
+        #
+        # Data chunks right of the partial
+        #
+        # start-at : last_partial_chunk_index + 1
+        # end-at   : last_data_chunk_index
+        if last_partial_chunk_index + 1 <= last_data_chunk_index:
+            nodes_right_of_partial = tree.get_subtree_proof_nodes(
+                chunk_index_to_path(last_partial_chunk_index + 1, self.path_bit_length),
+                last_data_chunk_path,
+            )
+        else:
+            nodes_right_of_partial = ()
+
+        #
+        # Padding
+        #
+        # start-at : last_partial_chunk_index + 1
+        # end-at   : N/A
+        if last_data_chunk_index + 1 <= self.sedes.chunk_count - 1:
+            first_padding_chunk_path = chunk_index_to_path(
+                last_data_chunk_index + 1, self.path_bit_length
+            )
+            actual_first_padding_node = tree.get_deepest_node_on_path(
+                first_padding_chunk_path
+            )
+            padding_nodes = tuple(
+                node
+                for node in tree.walk(actual_first_padding_node.path)
+                if node.is_terminal
+            )
+        else:
+            padding_nodes = ()
+
+        # Now re-assembly the sections of the tree for the minimal proof for
+        # the partial data.
+        partial_nodes = (
+            nodes_left_of_partial
+            + leaf_nodes_for_partial
+            + nodes_right_of_partial
+            + padding_nodes
+        )
+        partial_elements = tuple(
+            ProofElement(path=node.path, value=node.get_value())
+            for node in partial_nodes
+        )
+        proof = Proof(self.hash_tree_root, partial_elements, self.sedes)
+        validate_proof(proof)
+
+        return Proof(self.hash_tree_root, partial_elements, self.sedes)
```

Comment:
whoops, apparently didnt save the last thought:
nit: many lines of comments, but still a lot going on in `to_partial`. consider decomposing into smaller funcs in a future change set.

Reply:
Yeah, this one is long.  I actually have a bugfix lined up for this function in a later PR.  I'll take the time to merge it back down into this one because I think I'd have caught it if I'd taken the time to split this logic out and test the individual parts.  Basically, there are a few off-by-one errors hiding in here that don't show up until you have things like partial proofs that hit certain boundary conditions that aren't currently being tested.